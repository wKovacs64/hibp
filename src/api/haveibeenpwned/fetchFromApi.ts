import { Breach, Paste } from './types';
import axios from './axiosInstance';
import {
  BAD_REQUEST,
  UNAUTHORIZED,
  FORBIDDEN,
  NOT_FOUND,
  TOO_MANY_REQUESTS,
} from './responses';

export type ApiData =
  | Breach // breach
  | Breach[] // breachedaccount, breaches
  | Paste[] // pasteaccount
  | string[] // dataclasses
  | null; // most endpoints can return an empty response

const blockedWithRayId = (rayId: string): string =>
  `Request blocked, contact haveibeenpwned.com if this continues (Ray ID: ${rayId})`;

/**
 * Fetches data from the supplied API endpoint.
 *
 * HTTP status code 200 returns an Object (data found).
 * HTTP status code 404 returns null (no data found).
 * HTTP status code 400 throws an Error (bad request).
 * HTTP status code 401 throws an Error (unauthorized).
 * HTTP status code 403 throws an Error (forbidden).
 * HTTP status code 429 throws an Error (too many requests).
 *
 * @internal
 * @private
 * @param {string} endpoint the API endpoint to query
 * @param {object} [options] a configuration object
 * @param {string} [options.apiKey] an API key from
 * https://haveibeenpwned.com/API/Key
 * @param {string} [options.baseUrl] a custom base URL for the
 * haveibeenpwned.com API endpoints (default:
 * `https://haveibeenpwned.com/api/v3`)
 * @param {string} [options.userAgent] a custom string to send as the User-Agent
 * field in the request headers (default: `hibp <version>`)
 * @returns {Promise<ApiData>} a Promise which resolves to the data resulting
 * from the query (or null for 404 Not Found responses), or rejects with an
 * Error
 */
export default (
  endpoint: string,
  /* istanbul ignore next: no need to test default empty object */
  options: { apiKey?: string; baseUrl?: string; userAgent?: string } = {},
): Promise<ApiData> => {
  const { apiKey, baseUrl, userAgent } = options;

  const headers: {
    'HIBP-API-Key'?: string;
    'User-Agent'?: string;
  } = {};

  if (apiKey) {
    headers['HIBP-API-Key'] = apiKey;
  }
  if (userAgent) {
    headers['User-Agent'] = userAgent;
  }

  const requestConfig = Object.assign(
    {},
    baseUrl ? { baseURL: baseUrl } : {},
    Object.keys(headers).length > 0
      ? {
          headers,
        }
      : {},
  );

  return Promise.resolve(axios.get<ApiData>(endpoint, requestConfig))
    .then(res => res.data)
    .catch(err => {
      if (err.response) {
        switch (err.response.status) {
          case BAD_REQUEST.status:
            throw new Error(BAD_REQUEST.statusText);
          case UNAUTHORIZED.status:
            throw new Error(err.response.data.message);
          case FORBIDDEN.status: {
            const rayId =
              err.response.headers && err.response.headers['cf-ray'];
            if (rayId) {
              throw new Error(blockedWithRayId(rayId));
            }
            throw new Error(FORBIDDEN.statusText);
          }
          case NOT_FOUND.status:
            return null;
          case TOO_MANY_REQUESTS.status:
            throw new Error(err.response.data.message);
          default:
            throw new Error(err.response.statusText);
        }
      } else {
        throw err;
      }
    });
};
