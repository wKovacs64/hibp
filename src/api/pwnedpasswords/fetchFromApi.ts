import fetch from '../web-fetch';
import { BAD_REQUEST } from './responses';

/**
 * Fetches data from the supplied API endpoint.
 *
 * HTTP status code 200 returns plain text (data found).
 * HTTP status code 400 throws an Error (bad request).
 *
 * @internal
 * @private
 * @param {string} endpoint the API endpoint to query
 * @param {object} [options] a configuration object
 * @param {string} [options.baseUrl] a custom base URL for the
 * pwnedpasswords.com API endpoints (default: `https://api.pwnedpasswords.com`)
 * @param {string} [options.userAgent] a custom string to send as the User-Agent
 * field in the request headers (default: `hibp <version>`)
 * @returns {Promise<string>} a Promise which resolves to the data resulting
 * from the query, or rejects with an Error
 */
export function fetchFromApi(
  endpoint: string,
  {
    baseUrl = 'https://api.pwnedpasswords.com',
    userAgent,
  }: { baseUrl?: string; userAgent?: string } = {},
): Promise<string> {
  const config = Object.assign(
    {},
    userAgent
      ? {
          headers: {
            'User-Agent': userAgent,
          },
        }
      : {},
  );

  const url = `${baseUrl.replace(/\/$/g, '')}${endpoint}`;

  return fetch(url, config).then((res) => {
    if (res.ok) return res.text();

    if (res.status === BAD_REQUEST.status) {
      return res.text().then((text) => {
        throw new Error(text);
      });
    }

    throw new Error(res.statusText);
  });
}
