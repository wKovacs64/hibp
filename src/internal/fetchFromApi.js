import axios from './axiosInstance';
import {
  BAD_REQUEST,
  FORBIDDEN,
  NOT_FOUND,
  TOO_MANY_REQUESTS,
} from './responses';

/**
 * Fetches data from the supplied API endpoint.
 *
 * HTTP status code 200 returns an Object (data found).
 * HTTP status code 404 returns null (no data found).
 * HTTP status code 400 throws an Error (bad request).
 * HTTP status code 403 throws an Error (forbidden).
 * HTTP status code 429 throws an Error (too many requests).
 *
 * @private
 * @param {string} endpoint the API endpoint to query
 * @returns {Promise} a Promise which resolves to the data resulting from the
 * query (or null for 404 Not Found responses), or rejects with an Error
 */
export default endpoint =>
  Promise.resolve(axios.get(endpoint))
    .then(res => res.data)
    .catch((err) => {
      if (err.response) {
        switch (err.response.status) {
          case BAD_REQUEST.status:
            throw new Error(BAD_REQUEST.statusText);
          case FORBIDDEN.status:
            throw new Error(FORBIDDEN.statusText);
          case NOT_FOUND.status:
            return null;
          case TOO_MANY_REQUESTS.status:
            throw new Error(err.response.data);
          default:
            throw new Error(err.response.statusText);
        }
      } else {
        throw err;
      }
    });
