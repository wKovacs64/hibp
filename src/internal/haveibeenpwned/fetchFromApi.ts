import { Breach, Paste } from 'types/remote-api.d';
import axios from './axiosInstance';
import {
  BAD_REQUEST,
  FORBIDDEN,
  NOT_FOUND,
  TOO_MANY_REQUESTS,
} from './responses';

export type ApiData =
  | Breach // breach
  | Breach[] // breachedaccount, breaches
  | Paste[] // pasteaccount
  | string[] // dataclasses
  | null; // most endpoints can return an empty response

const blockedWithRayId = (rayId: string): string =>
  `Request blocked, contact haveibeenpwned.com if this continues (Ray ID: ${rayId})`;

/**
 * Fetches data from the supplied API endpoint.
 *
 * HTTP status code 200 returns an Object (data found).
 * HTTP status code 404 returns null (no data found).
 * HTTP status code 400 throws an Error (bad request).
 * HTTP status code 403 throws an Error (forbidden).
 * HTTP status code 429 throws an Error (too many requests).
 *
 * @internal
 * @private
 * @param {string} endpoint the API endpoint to query
 * @param {object} [options] a configuration object
 * @param {string} [options.userAgent] a custom string to send as the User-Agent
 * field in the request headers (default: `hibp <version>`)
 * @returns {Promise<ApiData>} a Promise which resolves to the data resulting
 * from the query (or null for 404 Not Found responses), or rejects with an
 * Error
 */
export default (
  endpoint: string,
  /* istanbul ignore next: no need to test default empty object */
  options: { userAgent?: string } = {},
): Promise<ApiData> => {
  const { userAgent } = options;

  const config = Object.assign(
    {},
    userAgent
      ? {
          headers: {
            'User-Agent': userAgent,
          },
        }
      : {},
  );

  return Promise.resolve(axios.get<ApiData>(endpoint, config))
    .then(res => res.data)
    .catch(err => {
      if (err.response) {
        switch (err.response.status) {
          case BAD_REQUEST.status:
            throw new Error(BAD_REQUEST.statusText);
          case FORBIDDEN.status: {
            const rayId =
              err.response.headers && err.response.headers['cf-ray'];
            if (rayId) {
              throw new Error(blockedWithRayId(rayId));
            }
            throw new Error(FORBIDDEN.statusText);
          }
          case NOT_FOUND.status:
            return null;
          case TOO_MANY_REQUESTS.status:
            throw new Error(err.response.data);
          default:
            throw new Error(err.response.statusText);
        }
      } else {
        throw err;
      }
    });
};
