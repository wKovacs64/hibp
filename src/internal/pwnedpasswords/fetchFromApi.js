import axios from './axiosInstance';
import { RANGE_BAD_REQUEST } from './responses';

/**
 * Fetches data from the supplied API endpoint.
 *
 * HTTP status code 200 returns plain text (data found).
 * HTTP status code 400 throws an Error (bad request).
 *
 * @private
 * @param {string} endpoint the API endpoint to query
 * @returns {Promise} a Promise which resolves to the data resulting from the
 * query, or rejects with an Error
 */
export default endpoint =>
  Promise.resolve(axios.get(endpoint))
    .then(res => res.data)
    .catch(err => {
      if (err.response) {
        switch (err.response.status) {
          case RANGE_BAD_REQUEST.status:
            throw new Error(err.response.data);
          default:
            throw new Error(err.response.statusText);
        }
      } else {
        throw err;
      }
    });
