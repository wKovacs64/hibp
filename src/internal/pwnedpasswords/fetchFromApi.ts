import axios from './axiosInstance';
import { BAD_REQUEST } from './responses';

/**
 * Fetches data from the supplied API endpoint.
 *
 * HTTP status code 200 returns plain text (data found).
 * HTTP status code 400 throws an Error (bad request).
 *
 * @internal
 * @private
 * @param {string} endpoint the API endpoint to query
 * @returns {Promise<string>} a Promise which resolves to the data resulting
 * from the query, or rejects with an Error
 */
export default (endpoint: string): Promise<string> =>
  Promise.resolve(axios.get<string>(endpoint))
    .then(res => res.data)
    .catch(err => {
      if (err.response) {
        switch (err.response.status) {
          case BAD_REQUEST.status:
            throw new Error(err.response.data);
          default:
            throw new Error(err.response.statusText);
        }
      } else {
        throw err;
      }
    });
