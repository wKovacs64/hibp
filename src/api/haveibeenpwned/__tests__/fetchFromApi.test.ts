import { mockFetch, mockResponse } from '../../../../test/utils';
import {
  OK,
  BAD_REQUEST,
  UNAUTHORIZED,
  FORBIDDEN,
  BLOCKED,
  TOO_MANY_REQUESTS,
} from '../responses';
import { fetchFromApi } from '../fetchFromApi';

describe('internal (haveibeenpwned): fetchFromApi', () => {
  const apiKey = 'my-api-key';

  describe('User-Agent', () => {
    // Node
    it('sets a User-Agent request header when outside the browser', () => {
      const originalNavigator = global.navigator;
      mockFetch.mockResolvedValueOnce(mockResponse({ status: OK.status }));

      global.navigator = undefined;
      return fetchFromApi('/service')
        .then(() => {
          expect(mockFetch).toHaveBeenCalledWith(expect.any(String), {
            headers: expect.objectContaining({
              'User-Agent': expect.any(String),
            }),
          });
        })
        .finally(() => {
          global.navigator = originalNavigator;
        });
    });

    // Browser
    it('does not set a User-Agent request header when inside the browser', () => {
      mockFetch.mockResolvedValueOnce(mockResponse({ status: OK.status }));

      // eslint-disable-next-line @typescript-eslint/consistent-type-assertions
      global.navigator = {} as Navigator;
      return fetchFromApi('/service').then(() => {
        expect(mockFetch).toHaveBeenCalledWith(expect.any(String), {
          headers: expect.not.objectContaining({
            'User-Agent': expect.any(String),
          }),
        });
      });
    });
  });

  describe('request failure', () => {
    it('re-throws request setup errors', () => {
      const ERR = new Error('Set sail for fail!');
      mockFetch.mockRejectedValueOnce(ERR);

      return expect(fetchFromApi('/service')).rejects.toEqual(ERR);
    });
  });

  describe('invalid account format', () => {
    it('throws a "Bad Request" error', () => {
      mockFetch.mockResolvedValueOnce(
        mockResponse({
          status: BAD_REQUEST.status,
          statusText: BAD_REQUEST.statusText,
        }),
      );

      return expect(
        fetchFromApi('/service/bad_request', { apiKey }),
      ).rejects.toMatchInlineSnapshot(
        `[Error: Bad request â€” the account does not comply with an acceptable format.]`,
      );
    });
  });

  describe('unauthorized', () => {
    it('throws an "Unauthorized" error', () => {
      mockFetch.mockResolvedValueOnce(
        mockResponse({
          status: UNAUTHORIZED.status,
          body: UNAUTHORIZED.body,
        }),
      );

      return expect(
        fetchFromApi('/service/unauthorized'),
      ).rejects.toMatchInlineSnapshot(
        `[Error: Access denied due to missing hibp-api-key.]`,
      );
    });
  });

  describe('forbidden request', () => {
    it('throws a "Forbidden" error if no cf-ray header is present', () => {
      mockFetch.mockResolvedValueOnce(
        mockResponse({
          status: FORBIDDEN.status,
          statusText: FORBIDDEN.statusText,
        }),
      );

      return expect(
        fetchFromApi('/service/forbidden', { apiKey }),
      ).rejects.toMatchInlineSnapshot(`[Error: Forbidden - access denied.]`);
    });

    it('throws a "Blocked Request" error if a cf-ray header is present', () => {
      mockFetch.mockResolvedValueOnce(
        mockResponse({
          headers: BLOCKED.headers,
          status: BLOCKED.status,
        }),
      );

      return expect(
        fetchFromApi('/service/blocked', { apiKey }),
      ).rejects.toMatchInlineSnapshot(
        `[Error: Request blocked, contact haveibeenpwned.com if this continues (Ray ID: someRayId)]`,
      );
    });
  });

  describe('rate limited', () => {
    it('throws a "Too Many Requests" error', () => {
      mockFetch.mockResolvedValueOnce(
        mockResponse({
          status: TOO_MANY_REQUESTS.status,
          body: TOO_MANY_REQUESTS.body,
        }),
      );

      return expect(
        fetchFromApi('/service/rate_limited', { apiKey }),
      ).rejects.toMatchInlineSnapshot(
        `[Error: Rate limit is exceeded. Try again in 2 seconds.]`,
      );
    });
  });

  describe('unexpected HTTP error', () => {
    it('throws an error with the response status text', () => {
      mockFetch.mockResolvedValueOnce(
        mockResponse({
          status: 999,
          statusText: 'Unknown - something unexpected happened.',
        }),
      );

      return expect(
        fetchFromApi('/service/unknown_response'),
      ).rejects.toMatchInlineSnapshot(
        `[Error: Unknown - something unexpected happened.]`,
      );
    });
  });

  describe('apiKey option', () => {
    it('is passed on as a request header', () => {
      mockFetch.mockResolvedValue(mockResponse({ status: OK.status }));

      return fetchFromApi('/service', { apiKey }).then(() => {
        expect(mockFetch).toHaveBeenCalledWith(expect.any(String), {
          headers: expect.objectContaining({ 'HIBP-API-Key': apiKey }),
        });
      });
    });
  });

  describe('userAgent option', () => {
    it('is passed on as a request header', () => {
      mockFetch.mockResolvedValue(mockResponse({ status: OK.status }));

      const ua = 'custom UA';

      return fetchFromApi('/service', { userAgent: ua }).then(() => {
        expect(mockFetch).toHaveBeenCalledWith(expect.any(String), {
          headers: expect.objectContaining({ 'User-Agent': ua }),
        });
      });
    });
  });

  describe('baseUrl option', () => {
    it('is used in the final URL', () => {
      mockFetch.mockResolvedValue(mockResponse({ status: OK.status }));

      const baseUrl = 'https://my-hibp-proxy:8080';
      const endpoint = '/service';

      return fetchFromApi(endpoint, { baseUrl }).then(() => {
        expect(mockFetch).toHaveBeenCalledWith(
          `${baseUrl}${endpoint}`,
          expect.any(Object),
        );
      });
    });
  });
});
