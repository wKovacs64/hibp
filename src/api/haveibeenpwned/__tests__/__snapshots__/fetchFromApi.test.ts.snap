// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`internal (haveibeenpwned): fetchFromApi forbidden request throws a "Blocked Request" error if a cf-ray header is present 1`] = `[Error: Request blocked, contact haveibeenpwned.com if this continues (Ray ID: someRayId)]`;

exports[`internal (haveibeenpwned): fetchFromApi forbidden request throws a "Forbidden" error if no cf-ray header is present 1`] = `[Error: Forbidden - access denied.]`;

exports[`internal (haveibeenpwned): fetchFromApi invalid account format throws a "Bad Request" error 1`] = `[Error: Bad request â€” the account does not comply with an acceptable format.]`;

exports[`internal (haveibeenpwned): fetchFromApi rate limited throws a "Too Many Requests" error 1`] = `[Error: Rate limit is exceeded. Try again in 2 seconds.]`;

exports[`internal (haveibeenpwned): fetchFromApi unauthorized throws an "Unauthorized" error 1`] = `[Error: Access denied due to missing hibp-api-key.]`;

exports[`internal (haveibeenpwned): fetchFromApi unexpected HTTP error throws an error with the response status text 1`] = `[Error: Unknown - something unexpected happened.]`;
